### PR TITLE
Read Token: Remove Exceptions

### PR DESCRIPTION
**Pull request by iNinja**
# Read Token: Remove Exceptions

- [x] You've read the [Contributor Guide](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/dev/Contributing.md) and [Code of Conduct](https://opensource.microsoft.com/codeofconduct/).
- [x] You've included unit or integration tests for your change, where applicable.
- [x] You've included inline docs for your change, where applicable.
- [x] If any gains or losses in performance are possible, you've included benchmarks for your changes. [More info](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/wiki/Benchmarking-in-Wilson)
- [x] There's an open issue for the PR that you are making. If you'd like to propose a new feature or change, please open an issue to discuss the change or find an existing issue.

## Description
- Implemented new version of ReadToken removing TokenValidationParameters, dropping the TokenReader delegate, and returning a TokenReadingResult that wraps provides ExceptionDetails without throwing, and saves its logs as MessageDetails.  
Actual creation of the JsonWebToken untouched at this point. Exceptions that are thrown as part of the token creation are wrapped and returned as part of the result.
- Added nullability annotations on new code.
- Added tests.

### CODE REVIEW

**Review by SaurabhMSFT — APPROVED** [review]

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ReadToken.cs [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+using TokenLogMessages = Microsoft.IdentityModel.Tokens.LogMessages;
+
+namespace Microsoft.IdentityModel.JsonWebTokens
+{
+    /// <summary>
+    /// Definition for the delegate that converts a string into an instance of <see cref="JsonWebToken"/>.
+    /// </summary>
+    /// <param name="token">A JSON Web Token (JWT) in JWS or JWE Compact Serialization format.</param>
+    /// <param name="callContext"></param>
+    /// <returns>A <see cref="IssuerValidationResult"/>that contains the results of validating the issuer.</returns>
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ReadToken.cs]: statement needs updated

File: src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs [hunk 2]
```diff
@@ -0,0 +1,83 @@
+﻿using System;
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs]: copyright header

File: src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+#nullable enable
+namespace Microsoft.IdentityModel.Tokens
+{
+    /// <summary>
+    /// Contains the result of reading that a <see cref="SecurityToken"/>.
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs]:
```suggestion
    /// Contains the result of reading a <see cref="SecurityToken"/>.
```

File: src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        public string? TokenInput { get; }
+
+        /// <summary>
+        /// Gets the <see cref="SecurityToken"/> that was read.
+        /// </summary>
+        public SecurityToken SecurityToken {
+            get
+            {
+                if (_securityToken is null)
+                    throw new InvalidOperationException("Attempted to retrieve the SecurityToken from a failed TokenReading result.");
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs]: generally it is an anti pattern to throw exceptions from property getters

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ReadTokenTests.cs [hunk 5]
```diff
@@ -0,0 +1,116 @@
+﻿using System;
```
Comment (keegan-caruso) [inline test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ReadTokenTests.cs]: copyright header

File: src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs [hunk 6]
```diff
@@ -0,0 +1,83 @@
+﻿using System;
+
+#nullable enable
+namespace Microsoft.IdentityModel.Tokens
+{
+    /// <summary>
+    /// Contains the result of reading that a <see cref="SecurityToken"/>.
+    /// The <see cref="TokenValidationResult"/> contains a collection of <see cref="ValidationResult"/> for each step in the token validation.
+    /// </summary>
+    internal class TokenReadingResult: ValidationResult
```
Comment (keegan-caruso) [inline src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs]:
```suggestion
    internal class TokenReadingResult : ValidationResult
```

File: src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs (hunk 2 above)
Comment (keegan-caruso) [reply src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs]: There is an editorconfig that should catch these...

File: test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ReadTokenTests.cs (hunk 5 above)
Comment (iNinja) [reply test/Microsoft.IdentityModel.JsonWebTokens.Tests/JsonWebTokenHandler.ReadTokenTests.cs]: Added.

File: src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs (hunk 4 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs]: Changed to method and documented the exception.

File: src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs (hunk 3 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs]: Updated.

File: src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs (hunk 2 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs]: Updated.

**Review by keegan-caruso — APPROVED** [review]

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ReadToken.cs (hunk 1 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.ReadToken.cs]: Updated.

File: src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs (hunk 6 above)
Comment (iNinja) [reply src/Microsoft.IdentityModel.Tokens/Validation/TokenReadingResult.cs]: Updated.

**Review by FuPingFranco — APPROVED** [review]

**Pull request merged**